lib/ASTGen/Sources/ASTGen/ASTGen.swift

```diff
@@ -13,35 +13,47 @@ extension UnsafePointer {
 }
 
 enum ASTNode {
-  case decl(UnsafeMutableRawPointer)
-  case stmt(UnsafeMutableRawPointer)
-  case expr(UnsafeMutableRawPointer)
-  case type(UnsafeMutableRawPointer)
-  case misc(UnsafeMutableRawPointer)
+  case decl(BridgedDecl)
+  case stmt(BridgedStmt)
+  case expr(BridgedExpr)
+  case type(BridgedTypeRepr)
+
+  var castToExpr: BridgedExpr {
+    guard case .expr(let bridged) = self else {
+      fatalError("Expected an expr")
+    }
+    return bridged
+  }
 
-  var rawValue: UnsafeMutableRawPointer {
-    switch self {
-    case .decl(let ptr):
-      return ptr
-    case .stmt(let ptr):
-      return ptr
-    case .expr(let ptr):
-      return ptr
-    case .type(let ptr):
-      return ptr
-    case .misc(let ptr):
-      return ptr
+  var castToStmt: BridgedStmt {
+    guard case .stmt(let bridged) = self else {
+      fatalError("Expected a stmt")
+    }
+    return bridged
+  }
+
+  var castToDecl: BridgedDecl {
+    guard case .decl(let bridged) = self else {
+      fatalError("Expected a decl")
     }
+    return bridged
+  }
+
+  var castToType: BridgedTypeRepr {
+    guard case .type(let bridged) = self else {
+      fatalError("Expected a type")
+    }
+    return bridged
   }
 
   var bridged: BridgedASTNode {
     switch self {
     case .expr(let e):
-      return BridgedASTNode(ptr: e, kind: .expr)
+      return BridgedASTNode(ptr: e.raw!, kind: .expr)
```

Comment:
Should we create a `BridgedASTNode` with a `nullptr` instead of crashing here? AFAICT nothing guarantees that `raw` is not `nil` here.

Reply:
`BridgedASTNode` has a non-nullable value, so we have to unwrap here. We could change `BridgedASTNode` to take a nullable value, but that would just be trading an explicit unwrap for an implicit null pointer access in C++ land. The alternative here would be to introduce separate nullable types for the base AST node wrapper types, e.g `BridgedExpr` and `BridgedNullableExpr`, and then parameters expecting a nullable expr could use the nullable version and you'd have to add an extra `.init(...)` when passing the argument. I can't say I'm a huge fan of that, though I do agree the force unwrap here isn't great.

Reply:
Or a completely different option: Do you think we actually need `ASTNode` or could we remove it altogether (definitely not this PR). To me `ASTNode` seems like an anti-pattern anyway because it generalizes too much. 

For example `ASTGen.generate(ExprSyntax)` should switch over all the expression nodes (for which we need to implement an `ExprSyntaxEnum`, which I think I have thought about adding for 2 years now) and then it could just return a `BridgedExprSyntax` without the need to go through `ASTNode`.

Reply:
The issue is for things like BraceStmt, where the elements truly are ASTNodes :/

Reply:
FWIW this is what it would look like with nullable wrappers ef7407804c87f2bd3826ab57c131fe73f8da24fe, I guess it's not _too_ bad? Still can't say I love it

Reply:
I think I prefer it with https://github.com/apple/swift/commit/ef7407804c87f2bd3826ab57c131fe73f8da24fe. At least it’s explicit about what it’s doing and the `asNullable` are annoying but not toooo bad. But it’s Friday evening and I might change my mind when I look at this again on Monday.

Reply:
Yeah, I don't really have particulary strong feelings either way, I've left it as-is for now, but am happy to do the nullable wrappers in a follow up

Reply:
Hmm the more I think about, the less I hate it, put up a PR: #69326